File: proyecto.java
// Proyecto.java

import java.util.ArrayList; // Importa la implementación dinámica de listas ArrayList.
import java.util.List;      // Importa la interfaz List para declarar colecciones.

 // Clase base Producto
class Producto {
    private int id;           // Identificador único del producto.
    private String nombre;    // Nombre del producto.
    private double precio;    // Precio unitario del producto.
    private int stock;        // Cantidad disponible en inventario.

    public Producto(int id, String nombre, double precio, int stock) { // Constructor que inicializa campos.
        this.id = id;                // Asigna el id recibido al atributo de la instancia.
        this.nombre = nombre;        // Asigna el nombre recibido al atributo de la instancia.
        this.precio = precio;        // Asigna el precio recibido al atributo de la instancia.
        this.stock = stock;          // Asigna el stock recibido al atributo de la instancia.
    }

    public void actualizarStock(int cantidad) { // Método para disminuir el stock (por ejemplo al vender).
        this.stock -= cantidad;                 // Resta la cantidad indicada al stock actual.
    }

    public String mostrarInfo() {               // Método que devuelve una representación legible del producto.
        return nombre + " - Precio: $" + precio + ", Stock: " + stock; // Construye y retorna la cadena.
    }

    public double getPrecio() { // Getter para el precio.
        return precio;          // Devuelve el precio del producto.
    }

    public String getNombre() { // Getter para el nombre.
        return nombre;          // Devuelve el nombre del producto.
    }

    public int getStock() {     // Getter para el stock.
        return stock;           // Devuelve la cantidad en inventario.
    }
}

 // Clase ProductoDigital que hereda de Producto
class ProductoDigital extends Producto {
    private String formato;     // Almacena el formato del producto digital (ej. PDF, MP3).

    public ProductoDigital(int id, String nombre, double precio, String formato) { // Constructor.
        super(id, nombre, precio, 0); // Llama al constructor padre; pone stock = 0 porque no aplica.
        this.formato = formato;       // Asigna el formato recibido al atributo.
    }

    @Override
    public String mostrarInfo() { // Sobrescribe mostrarInfo para incluir formato y obviar stock.
        return getNombre() + " (Digital - " + formato + ") - Precio: $" + getPrecio(); // Construye y retorna la cadena.
    }
}

 // Clase Cliente
class Cliente {
    private int id;         // Identificador único del cliente.
    private String nombre;  // Nombre del cliente.
    private String correo;  // Correo electrónico del cliente.

    public Cliente(int id, String nombre, String correo) { // Constructor.
        this.id = id;            // Asigna el id.
        this.nombre = nombre;    // Asigna el nombre.
        this.correo = correo;    // Asigna el correo.
    }

    public String registrarse() { // Método que simula el registro y devuelve un mensaje.
        return "Cliente " + nombre + " registrado con correo " + correo; // Mensaje de confirmación.
    }

    public Pedido realizarPedido(Carrito carrito) { // Crea un pedido a partir del carrito proporcionado.
        return new Pedido(9001, this, carrito, "Pendiente"); // Devuelve un Pedido con estado inicial "Pendiente".
    }

    public String getNombre() { // Getter para el nombre (útil para mensajes).
        return nombre;          // Devuelve el nombre del cliente.
    }
}

 // Clase Carrito
class Carrito {
    private int id;                          // Identificador del carrito.
    private List<ItemCarrito> productos;     // Lista que guarda objetos ItemCarrito (producto + cantidad).

    public Carrito(int id) { // Constructor.
        this.id = id;                         // Asigna el id del carrito.
        this.productos = new ArrayList<>();   // Inicializa la lista de productos vacía.
    }

    public void agregarProducto(Producto producto, int cantidad) { // Agrega un producto y su cantidad al carrito.
        productos.add(new ItemCarrito(producto, cantidad)); // Añade un nuevo ItemCarrito a la lista.
        if (!(producto instanceof ProductoDigital)) {       // Si el producto NO es digital (es físico)...
            producto.actualizarStock(cantidad);             // ...descuenta esa cantidad del stock del producto.
        }
    }

    public double calcularTotal() { // Calcula el total del carrito multiplicando precio x cantidad.
        double total = 0;           // Inicializa acumulador en 0.
        for (ItemCarrito item : productos) { // Recorre cada elemento del carrito.
            total += item.getProducto().getPrecio() * item.getCantidad(); // Suma precio * cantidad al total.
        }
        return total; // Devuelve el total acumulado.
    }

    public List<ItemCarrito> getProductos() { // Getter para obtener la lista de productos del carrito.
        return productos;                      // Retorna la lista de ItemCarrito.
    }
}

 // Clase auxiliar para manejar producto + cantidad dentro del carrito
class ItemCarrito {
    private Producto producto; // Referencia al producto.
    private int cantidad;      // Cantidad de ese producto en el carrito.

    public ItemCarrito(Producto producto, int cantidad) { // Constructor.
        this.producto = producto; // Asigna el producto.
        this.cantidad = cantidad; // Asigna la cantidad.
    }

    public Producto getProducto() { // Getter del producto.
        return producto;            // Devuelve el producto asociado.
    }

    public int getCantidad() { // Getter de la cantidad.
        return cantidad;       // Devuelve la cantidad almacenada.
    }
}

 // Clase Pedido
class Pedido {
    private int id;         // Identificador del pedido.
    private Cliente cliente;// Cliente que hizo el pedido.
    private Carrito carrito; // Carrito asociado al pedido.
    private String estado;   // Estado actual del pedido (p.ej., "Pendiente", "Procesado").

    public Pedido(int id, Cliente cliente, Carrito carrito, String estado) { // Constructor.
        this.id = id;             // Asigna el id del pedido.
        this.cliente = cliente;   // Asigna la referencia al cliente.
        this.carrito = carrito;   // Asigna la referencia al carrito.
        this.estado = estado;     // Asigna el estado inicial.
    }

    public String procesarPedido() { // Método que procesa el pedido y cambia su estado a "Procesado".
        this.estado = "Procesado"; // Actualiza el estado del pedido.
        return "Pedido " + id + " procesado para cliente " + cliente.getNombre() + "."; // Retorna mensaje de confirmación.
    }

    public void cambiarEstado(String nuevoEstado) { // Permite cambiar el estado a cualquier valor.
        this.estado = nuevoEstado; // Asigna el nuevo estado pasado como argumento.
    }

    public String getEstado() { // Getter para consultar el estado actual del pedido.
        return estado;         // Devuelve el estado.
    }
}

 // Clase principal con el método main que actúa como punto de entrada de la aplicación
public class Proyecto {
    public static void main(String[] args) { // Método main: se ejecuta al iniciar la aplicación.
        Producto p1 = new Producto(101, "Laptop Lenovo", 2500.0, 5); // Crea un producto físico con id 101 y stock 5.
        Producto p2 = new Producto(102, "Mouse Logitech", 50.0, 20); // Crea otro producto físico con id 102 y stock 20.
        Producto p3 = new ProductoDigital(103, "Ebook Python", 15.0, "PDF"); // Crea un producto digital (formato PDF).

        Cliente cliente1 = new Cliente(1, "Ana Torres", "ana@email.com"); // Instancia un cliente con datos.
        Carrito carrito1 = new Carrito(501); // Crea un carrito con id 501.

        carrito1.agregarProducto(p1, 1); // Agrega 1 unidad de la laptop al carrito (reduce stock de p1 a 4).
        carrito1.agregarProducto(p2, 2); // Agrega 2 unidades del mouse (reduce stock de p2 a 18).
        carrito1.agregarProducto(p3, 1); // Agrega 1 ebook al carrito (no afecta stock porque es digital).

        Pedido pedido1 = cliente1.realizarPedido(carrito1); // El cliente genera un pedido a partir del carrito.

        // Salidas en consola para verificar funcionamiento
        System.out.println(cliente1.registrarse()); // Imprime el mensaje de registro del cliente.

        System.out.println("\nProductos en el carrito:"); // Encabezado de la lista de productos.
        for (ItemCarrito item : carrito1.getProductos()) { // Itera por cada ItemCarrito en el carrito.
            System.out.println("- " + item.getProducto().mostrarInfo() + " x" + item.getCantidad()); // Imprime info y cantidad.
        }

        System.out.println("\nTotal del carrito: " + carrito1.calcularTotal()); // Imprime el total a pagar.
        System.out.println(pedido1.procesarPedido()); // Procesa el pedido e imprime la confirmación.
        System.out.println("Estado final del pedido: " + pedido1.getEstado()); // Imprime el estado final del pedido.
    }
}

File: proyecto.py
class Producto:  # Define la clase base para productos físicos.
    def __init__(self, id, nombre, precio, stock):  # Constructor con identificador, nombre, precio y stock.
        self.id = id               # Asigna el identificador del producto.
        self.nombre = nombre       # Asigna el nombre del producto.               #SELF es una referencia al objeto actual de la clase.
        self.precio = precio       # Asigna el precio del producto.
        self.stock = stock         # Asigna la cantidad disponible en inventario.

    def actualizar_stock(self, cantidad):  # Método para decrementar el stock según una cantidad vendida.
        self.stock -= cantidad             # Resta la cantidad al stock actual.

    def mostrar_info(self):  # Devuelve una cadena con la información del producto.
        return f"{self.nombre} - Precio: ${self.precio}, Stock: {self.stock}"  # Formatea y retorna la info.


class ProductoDigital(Producto):  # Clase para productos digitales que hereda de Producto.
    def __init__(self, id, nombre, precio, formato):  # Constructor específico para productos digitales.
        super().__init__(id, nombre, precio, stock=None)  # Llama al constructor padre, sin stock (no aplica).
        self.formato = formato     # Guarda el formato del archivo (por ejemplo, PDF, MP3, etc.).

    def mostrar_info(self):  # Sobrescribe la presentación de info para productos digitales.
        return f"{self.nombre} (Digital - {self.formato}) - Precio: ${self.precio}"  # Muestra sin stock.


class Cliente:  # Representa a un cliente del sistema.
    def __init__(self, id, nombre, correo):  # Constructor con datos básicos del cliente.
        self.id = id               # Identificador único del cliente.
        self.nombre = nombre       # Nombre del cliente.
        self.correo = correo       # Correo electrónico del cliente.

    def registrarse(self):  # Simula el registro del cliente, devolviendo un mensaje.
        return f"Cliente {self.nombre} registrado con correo {self.correo}"  # Mensaje de confirmación.

    def realizar_pedido(self, carrito):  # Crea un pedido a partir del carrito asociado.
        pedido = Pedido(9001, self, carrito, "Pendiente")  # Instancia un Pedido con estado inicial.
        return pedido              # Retorna el pedido creado.


class Carrito:  # Modelo de un carrito de compras.
    def __init__(self, id):  # Constructor del carrito.
        self.id = id                  # Identificador del carrito.
        self.productos = []           # Lista de tuplas (producto, cantidad) agregados al carrito.

    def agregar_producto(self, producto, cantidad=1):  # Agrega un producto con una cantidad al carrito.
        self.productos.append((producto, cantidad))    # Añade la tupla (producto, cantidad) a la lista.
        if isinstance(producto, Producto) and not isinstance(producto, ProductoDigital):  # Si es físico...
            producto.actualizar_stock(cantidad)        # ...disminuye el stock del producto físico.

    def calcular_total(self):  # Calcula el total a pagar por los productos del carrito.
        total = 0              # Inicializa el acumulador del total.
        for producto, cantidad in self.productos:  # Recorre cada tupla (producto, cantidad).
            total += producto.precio * cantidad    # Suma precio * cantidad al total.
        return total            # Retorna el total calculado.


class Pedido:  # Representa un pedido realizado por un cliente.
    def __init__(self, id, cliente, carrito, estado):  # Constructor del pedido.
        self.id = id                 # Identificador del pedido.
        self.cliente = cliente       # Referencia al cliente que realiza el pedido.
        self.carrito = carrito       # Referencia al carrito asociado al pedido.
        self.estado = estado         # Estado actual del pedido (p.ej., Pendiente, Procesado).

    def procesar_pedido(self):  # Cambia el estado del pedido a Procesado y devuelve un mensaje.
        self.estado = "Procesado"  # Actualiza el estado interno del pedido.
        return f"Pedido {self.id} procesado para cliente {self.cliente.nombre}."  # Mensaje de confirmación.

    def cambiar_estado(self, nuevo_estado):  # Permite actualizar el estado del pedido arbitrariamente.
        self.estado = nuevo_estado           # Asigna el nuevo estado.


if __name__ == "__main__":  # Punto de entrada: se ejecuta solo si el archivo se corre directamente.
    p1 = Producto(101, "Laptop Lenovo", 2500.0, 5)       # Crea un producto físico con stock 5.
    p2 = Producto(102, "Mouse Logitech", 50.0, 20)       # Crea otro producto físico con stock 20.
    p3 = ProductoDigital(103, "Ebook Python", 15.0, "PDF")  # Crea un producto digital en formato PDF.

    cliente1 = Cliente(1, "Ana Torres", "ana@email.com")  # Instancia un cliente con sus datos.
    carrito1 = Carrito(501)                               # Crea un carrito de compras.

    carrito1.agregar_producto(p1, 1)  # Agrega 1 unidad de la laptop al carrito (stock de p1 baja a 4).
    carrito1.agregar_producto(p2, 2)  # Agrega 2 unidades del mouse (stock de p2 baja a 18).
    carrito1.agregar_producto(p3, 1)  # Agrega 1 ebook (no afecta stock por ser digital).

    pedido1 = cliente1.realizar_pedido(carrito1)  # El cliente genera un pedido con el carrito actual.

    print(cliente1.registrarse())  # Imprime el mensaje de registro del cliente.
    print("\nProductos en el carrito:")  # Título para listar productos del carrito.
    for producto, _ in carrito1.productos:  # Itera por los productos en el carrito (ignora cantidad aquí).
        print("-", producto.mostrar_info())  # Imprime la información de cada producto.

    print(f"\nTotal del carrito: {carrito1.calcular_total()}")  # Muestra el total a pagar.

    print(pedido1.procesar_pedido())       # Procesa el pedido e imprime el mensaje de confirmación.
    print("Estado final del pedido:", pedido1.estado)  # Muestra el estado final del pedido.
